3.2Expand Regex DoS check to include String#match and #match? coercion

bensheldon opened the pull request:
Closes #1714.

I added additional behavior to the existing `check_regex_dos.rb` file. I thought it seemed like the correct location because

bensheldon commented:
oops, sorry this got submitted in a half-written state. Cat on keyboard problems 😸

presidentbeef commented on lib/brakeman/checks/check_regex_dos.rb:
```diff
@@ -11,12 +11,20 @@ class Brakeman::CheckRegexDoS < Brakeman::BaseCheck
     Brakeman.debug "Finding dynamic regexes"
     calls = tracker.find_call :method => [:brakeman_regex_interp]
+    COERCES_STRING_TO_REGEX.each do |coercion_method|
+      calls.concat tracker.find_call(:method => [coercion_method]).select { |call| string?(call[:target]) }
```
Could simplify this to

```ruby
tracker.find_call(:methods => CEORCES_STRING_TO_REGEX).select { ... }
```

bensheldon commented:
@presidentbeef I updated the code based on your feedback. Thank you! 🙏🏻 

I also made one behavior change: I added a method called `string_or_modified_string?` that recurses down the target to see if the initial call was a String. I added this because I found making a trivial modification like `downcase` would not trigger the check (e.g. `"haystack".downcase.match(params[:unsafe])`). I couldn't find an existing recursive method, but maybe I overlooked one.

presidentbeef commented:
In real code, would we expect to see string literals like the test cases? :thinking:

bensheldon commented:
> In real code, would we expect to see string literals like the test cases? 🤔

This was the code we discovered (lightly anonymized):

```ruby
result = @active_record_resource.a_hash_value
  .keys
  .select { |key| key.downcase.match?(unsafe_string_query) }
  .sort
```

I'd love advice for better ways to identify `[String]#match?([String])`.
